server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardSegmentsCommand.java

```diff
@@ -0,0 +1,586 @@
+                    }
+                }
+                return null;
+            },
+            () -> new ElasticsearchException("Unable to resolve shard path for index [" + indexName + "]"));
+    }
+
+    /**
+     * resolve {@code ShardPath} based on a full index or translog path
+     * @param environment
+     * @param path
+     * @return
+     * @throws IOException
+     * @throws UserException
+     */
+    private ShardPath resolveShardPath(Environment environment, Path path) throws IOException, UserException {
+        final Settings settings = environment.settings();
+        final Path shardParentPath = path.getParent();
+        final Path indexPath = path.resolve(ShardPath.INDEX_FOLDER_NAME);
+        if (Files.exists(indexPath) == false || Files.isDirectory(indexPath) == false) {
+            throw new ElasticsearchException("index directory [" + indexPath + "], must exist and be a directory");
+        }
+
+        final IndexMetaData indexMetaData =
+            IndexMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, shardParentPath);
+        final Index index = indexMetaData.getIndex();
+        final ShardId shardId;
+        final String fileName = path.getFileName().toString();
+        if (Files.isDirectory(path) && fileName.chars().allMatch(Character::isDigit)) {
+            int id = Integer.parseInt(fileName);
+            shardId = new ShardId(index, id);
+        } else {
+            throw new ElasticsearchException("Unable to resolve shard id from " + path.toString());
+        }
+
+        final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings);
+
+        return resolveShardPath(environment,
+            (nodeLockId, nodePath) -> {
+                final Path shardPathLocation = nodePath.resolve(shardId);
+                final ShardPath shardPath = ShardPath.loadShardPath(logger, shardId, indexSettings, new Path[]{shardPathLocation},
+                    nodeLockId, nodePath.path);
+                if (shardPath != null && shardPath.resolveIndex().equals(indexPath)) {
+                    return shardPath;
+                }
+                return null;
+            },
+            () -> new ElasticsearchException("Unable to resolve shard path for path " + path.toString()));
+    }
+
+    private ShardPath resolveShardPath(final Environment environment,
+                                       final CheckedBiFunction<Integer, NodeEnvironment.NodePath, ShardPath, IOException> function,
+                                       final Supplier<ElasticsearchException> exceptionSupplier)
+        throws IOException, UserException {
+        // have to iterate over possibleLockId as NodeEnvironment - but fail if node owns lock
+        final Settings settings = environment.settings();
+        final Path[] dataFiles = environment.dataFiles();
+        // try to resolve shard path in case of multi-node layout per environment (hope it's only integration tests)
```

Comment:
Hope?

Reply:
Dropped confusing part of comment, adding test that covers multi-node layout per environment